### ISSUE TITLE
Improve functioning of ScErrInputQueueFull

### ISSUE THREAD

**Issue by dan31** [Starcounter/Starcounter#3388]
ScErrInputQueueFull is so far our anti-champion:

https://github.com/Starcounter/Starcounter/issues/2565
https://github.com/Starcounter/Starcounter/issues/2900
https://github.com/Starcounter/Starcounter/issues/2646
https://github.com/Starcounter/Starcounter/issues/1305
https://github.com/Starcounter/Starcounter/issues/2912
https://github.com/Starcounter/Starcounter/issues/1748
https://github.com/Starcounter/Starcounter/issues/1293
https://github.com/Starcounter/Starcounter/issues/950

Currently the problem with this queue is formulated as follows:
> We have a very strange issue on one of our server. All the handlers are not responding, nor the SQL queries are executing in SC Admin Panel. I attached the debugger, placed a breakpoint in a handler, went to that handler in browser. The code control never enters in that handler. and the apps crashed after a while. System.Exception: ScErrInputQueueFull (SCERR1026): Operation failed because the target input queue was full.

Here I cite @atifsheikh. Me and @linkdata experienced the same recently while stress-testing LogStreamer. After running for a while, the instance starts to stand still, SQL queries from administrator take infinite time to complete, no handlers are responding (takes infinite time to respond), and you can only kill an instance to resolve the situation. Counting shows that the number of tasks in a queue has grown by that moment (@linkdata, how idid you figure that out? since we actually got no exception in a log).

Of course, in all such cases, it is simply that too many tasks have been scheduled at a time. However, while native .NET TaskScheduler manage this by running multiple tasks on a limited number of threads, so that you can queue more and more tasks (killing performance with syncronisation context switching if not doing this run properly and/or not utilising ```async/await``` pattern), in Starcounter "something strange" happens instead.

Point of this issue is not to change the (currently kernel-dependent) scheduling mechanism. This will be addressed in future releases of 2.x on ```pnext```, when @linkdata moves us to his new scheduler. This will also let us in the path to using all native .NET scheduling along the way and also ```async/await```.

Instead, in here we improve the Starcounter's task scheduling on a managed level so that this situation is no more wierd, in other words we need to clarify behaviour in a corner case. We need to understand what happens in this "weird" situation and address it with the least possible effort. We will also add docs to .io explaining how not to shoot yourself with ```.ScheduleTask``` in Starcounter.

I will try to find time to investigate further on during this week, hence assigning to myself as of now.

**Comment by dan31** [comment]
CC @alemoi @chrhol

**Comment by linkdata** [comment]
> (@linkdata, how idid you figure that out? since we actually got no exception in a log)

The `LogStreamer` testing started behaving strange when we added `Bulker` to the mix, so I looked at that fairly soon in the troubleshooting process. I added semaphores around it's use of `ScheduleTask()` just to eliminate it as a source of the problems, and the systems immediately stabilized.

In writing the `LogStreamer`, using semaphores and queues to manage the work was part of the design from the start (that's simply how you write asynchronous programs), so it never elicited the "weird" behavior from Starcounter.

**Comment by dan31** [comment]
We took a deliberate design session last 2 days and here what we come up with.

1. __Queue becomes no longer static__. The unmanaged code in our old scheduler for a fixed circular queue is to be redone into dynamically growing list-like queue. To be done by @alemoi and @dan31. @alemoi - short note: before operation, we have to wrap it with a microbenchmark to see how we affect performance with our changes.

2. In order to let app developers manage pressure on the system on their own, using __their proper queueing__ which depends on their subject domain, the queue in Starcounter will expose metrics. E.g., number of items currently in a queue (per scheduler), or queue's consumption rate. Using this input, app developers will be able to better handle pressure control.

3. After a fix, a __deep article on Starcounter db scheduling and Starcounter db jobs queuing will be created__. In addition, __we will explain there to our users that they have to manage pressure on the system properly__, and we will show to them how to do it in C#. This one has nothing to do with Starcounter itself, but it is hard to expect a knowledge of concurrent programming from an average Starcounter user. Hence we need to help them understand the basics better. We will show probably simple yet realistic examples with patterns like: persisting queues in Starcounter, using ConcurrentQueue, slim semaphores.

4. We __do not make ScheduleTask return .NET Task__. The reason being is that currently it will not work with async/await in C# in Starcounter. It will only work after we move to a new scheduler and ASP.NET core codehost (or kestrel-married .NET 4.6 codehost) together with it. However, seeing a Task returned, people would try to await on it for sure. Hence we won't break the least surprise principle.

CC @alemoi @eriohl @linkdata @bigwad @Starcounter-Jack

**Comment by linkdata** [comment]
The [page linked to by the error](https://github.com/Starcounter/Starcounter/wiki/SCERR1026) now contains sample code for a wrapper around `Scheduler.ScheduleTask()` that will catch the exceptions and retry for a given amount of time, as well as links to sample projects which do producer-consumer queue management.

**Comment by alemoi** [comment]
I made a few performance fixes and added queues on top for good distribution. So now its no longer possible to get the ScErrInputQueueFull exception. Distribution and performance improved a lot. A set of automatic tests added, here is an example of performance:
```
[21:19:07][SimpleIndependentTests_hosted_test-exe] -- Running process "star C:\BuildAgent\work\sc-develop-daily-1245\Level1\src\Tests\SimpleIndependentTests\bin\Release\SimpleIndependentTests_hosted_test.exe"
[21:19:37][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SimplestAsyncTaskOnAnyScheduler_PerSec' value='875196']
[21:19:40][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SimplestAsyncTaskOnSpecificScheduler1_PerSec' value='2913752']
[21:19:44][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SimplestAsyncTaskOnSpecificScheduler0_PerSec' value='2397506']
[21:19:52][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SmallAsyncTaskOnAnyScheduler_PerSec' value='1495662']
[21:20:17][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SmallAsyncTaskOnSpecificScheduler1_PerSec' value='384186']
[21:20:42][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SmallAsyncTaskOnSpecificScheduler0_PerSec' value='390960']
[21:20:53][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SimplestSyncTaskOnAnyScheduler_PerSec' value='97200']
[21:21:03][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SimplestSyncTaskOnSpecificScheduler1_PerSec' value='92850']
[21:21:04][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='SimplestSyncTaskOnSpecificScheduler0_PerSec' value='25000000']
[21:21:09][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='LargeAsyncTaskOnAnyScheduler_PerSec' value='35511']
[21:21:42][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='LargeAsyncTaskOnSpecificScheduler1_PerSec' value='6035']
[21:22:15][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='LargeAsyncTaskOnSpecificScheduler0_PerSec' value='6145']
[21:22:51][SimpleIndependentTests_hosted_test-exe] ##teamcity[buildStatisticValue key='LargeSyncTaskOnAnyScheduler_PerSec' value='5564']
```

The documentation is updated as well.

**Comment by alemoi** [comment]
Forgot to mention that existing code that has "private" workarounds for queue limit - should probably be cleaned now: @andwah @linkdata you maybe have some.

**Comment by andwah** [comment]
@alemoi  Nice work. we all was waiting for a neat solution.

Dont think i have any "private" workarounds as far as i can remember.

**Comment by dan31** [comment]
Great fix long waited :) Not closing it yet, let's see how it survives the staging. However, @alemoi added the new set of tests which reveal a very promising situation.

**Comment by bigwad** [comment]
I see several problems with this implementation:
##### 1. Major: async task may never gets executed. 
Due to race condition between scheduler queue consumer and producer we may reach the code at DbSession.cs:217 having asyncTasksPerScheduler_[schedId].IsEmpty==true. Thus we are about to add a task to the managed scheduler queue and do not signal unmanaged one.
The same effect if we reach DbSession.cs:275 having unmanaged queue empty.

##### 2. Medium: async tasks assigned for the same scheduler may get executed not in the same order as they were scheduled.
I'm not sure how critical it is, but it definitely differs from previous experience. 

##### 3. Medium: deadlock maybe introduced due to infinite retry at DbSession.cs:356. 
Again, this is very specific situation and I'm not sure how critical it is, but it differs from previous experience.

**Comment by alemoi** [comment]
Thanks for the great input!
Regarding the first one, yes, it was addressed in my first implementation but I forgot to fix it in latter variant. Good catch! Regarding the 2 and 3 I will take a look, or you can explain how out of ordering is happening.

**Comment by bigwad** [comment]
> how out of ordering is happening

Imagine the size of unmanaged queue is N. When you schedule N+1 tasks, first N tasks go to unmanaged queue and the last one goes to managed. Then the first unmanaged tasks gets executed and right after it finishes scheduler will run task from the managed queue, not letting other tasks from unmanaged go first.

**Comment by linkdata** [comment]
I've written an evil little test program to help @alemoi out with testing: 
https://github.com/StarcounterSamples/TaskSpammer

**Comment by linkdata** [comment]
`TaskSpammer` functionality:
* Overloads the task queues by producing more work than is consumed.
* Checks that tasks are being run on the scheduler that was requested.
* Checks that tasks on the same scheduler are executed in order.
* Checks that tasks on the same scheduler are not executed in [parallel (concurrent is OK)](http://stackoverflow.com/questions/1897993/difference-between-concurrent-programming-and-parallel-programming).

**Comment by dan31** [comment]
@linkdata nice! Any evil materialised with a TaskSpammer so far?? :)

**Comment by linkdata** [comment]
Both out-of-order and in-parallel asserts trigger. @alemoi is aware and working on it.

**Comment by bigwad** [comment]
I've take a look at the test. 
Regarding parallel execution test fails since it's not aligned with scheduler's design. Actually, the only thing scheduler serializes is the access to ThreadData.ContextHandle to ensure nonconcurrent access to the kernel. In other words, once a task leaves any wait function, it could be executed in parallel with other tasks on the same scheduler up to the next access to ThreadData.ContextHandle. Thus, scheduler gives no general warranty on mutual exclusion of tasks. It only ensure correctness of kernel access. So, no fix from @alemoi required, but I think, it should be mentioned in docs.
Considering this thing with tasks could be executed in parallel, out-of-order issue become even less severe.

To fix the test it's enough to put empty transaction at the start of Lock function:
```
static void Lock(byte schedId) {
  Db.Transact(() => {
  });
...
```

**Comment by linkdata** [comment]
Actually, I don't think we make any concurrency or parallelism guarantees at all. We guarantee that the task runs on a SC scheduler and you can do transactions, but that's about it. I'll change `TaskSpammer` to test for that instead.

I do think the start order within a single scheduler is worth guaranteeing if we can do so, so I will leave those tests in for now.

**Comment by linkdata** [comment]
I removed the concurrency checks from `TaskSpammer` and added checks that `Db.Transact()` is available, threadsafe and runs on the correct scheduler.

**Comment by alemoi** [comment]
@bigwad again thank you for checking this!

> Thus, scheduler gives no general warranty on mutual exclusion of tasks. It only ensure correctness of kernel access.

I actually thought that one scheduler can't run two tasks in parallel, only one at a time, i.e. cooperatively (not only the database access part, but the whole scheduled task). @eriohl can you please confirm @bigwad observation?

**Comment by eriohl** [comment]
A single scheduler can run multiple threads in parallell. Only one will be "attached" (capable of accessing the kernel) at a time however. The other ones will either be detached or blocked. The detached ones can still run and do stuff other then acessing the kernel (or the query processor).

A scheduler can of course also switch between tasks (start task 2 before completing task 1). But this is in parallell in the same way as a single-threaded CPU can run multiple threads.

**Comment by alemoi** [comment]
@eriohl thank you for clarification!

I just pushed the fixes for point 1 and 2. The performance is approximately the same. After running current version of TaskSpammer - no errors were observed.

**Comment by alemoi** [comment]
Note that in point 1 fix I assume that the size of the queue is more than the number of schedulers (because SYNC tasks can theoretically pollute the queue if it would be too short).
Regarding point 3 I'm not sure if its applicable, because of `Thread.Sleep` the scheduler will switch to a new task (meaning that it would be picked from the queue) and then get back to the previous that was placed aside. So even if the queue is of size 1 - it should work. There is also a requirement that users should not get any exceptions that the queue is full. We can try, lets say for 10 seconds, and then through an exception, but I suppose this is not desired, right @dan31 ?

**Comment by dan31** [comment]
>  We can try, lets say for 10 seconds, and then through an exception, but I suppose this is not desired, right @dan31 ?

On one hand, a good practice is to introduce timeouts whenever possible. "Wait forever" situation leads to disasters in production, and there "circuit breaker" is a typical remedy. I believe you are thinking in a right direction. However, when exactly would this happen? Is it somewhat rare situation that might however take place and then "breaking the circuit" will help resolve it, or this will be happening frequently? To help you on, I don't have enough info to make further judgments.

**Comment by linkdata** [comment]
I wasn't happy with the load (or rather, lack thereof) that `TaskSpammer` was producing, so I rewrote it to use threads instead of timers. The good news is that it worked a treat - it can now push really hard. The bad news is that it occasionally finds an out-of-order execution.

You can now leave it running for extended periods of time, and it will strive to maintain a constant load on the scheduling queues. On my machine, leaving it running for a minute or two is usually enough for the race condition to trigger.

**Comment by linkdata** [comment]
I think this race condition may be because of OS thread scheduling though - it could be that SC is starting them in correct order but on different threads, which means it's up to the OS which thread actually manages to run the sequence number checking code first.

**Comment by alemoi** [comment]
> The good news is that it worked a treat - it can now push really hard. The bad news is that it occasionally finds an out-of-order execution.

For executing async tasks on a specific scheduler there is only one queue now, I don't use internal queue anymore, so everything should be in order unless, as you say: 

> it could be that SC is starting them in correct order but on different threads, which means it's up to the OS which thread actually manages to run the sequence number checking code first.

**Comment by alemoi** [comment]
@eriohl thank you for clarification! I am quite concerned about parallel execution on the same scheduler. Is there a practical reason for this? When exactly can it occur? I am asking because we have a lot of code that relies on the fact that two tasks can not run in parallel on the same scheduler, for example, for sessions and viewmodels (when, for example, user schedules two tasks on the same scheduler to do some changes on the viewmodel and push them). Is it possible to turn the parallel execution off or its deep in design?

**Comment by alemoi** [comment]
@bigwad provided a nice explanation on how scheduling works and why there is parallel execution possible (and actually unavoidable). I think at this point we can close this and corresponding issues, until something pops up related to tasks queue. What do you think @dan31 ?
